Setup mods layout

ShouvikGhosh2048 opened the pull request:
Closes #530.

ShouvikGhosh2048 commented:
@otobot1

Still a work in progress but you can try it out and mention any changes you'd like.
<img width="830" alt="mods-page-1" src="https://github.com/celestemods-com/CelesteMods/assets/91585022/6f5c5a2f-d7ce-4f46-8796-9672b65425d1">
<img width="847" alt="mods-page-2" src="https://github.com/celestemods-com/CelesteMods/assets/91585022/500da63b-02a2-4ed7-a148-eca3cd276c0a">

What values does the `rating popover` in the mockup use? Mods don't have difficulty/quality.

otobot1 commented:
Yeah the `rating popover` is a mistake in the mockup that didn't get fixed - the designer got busy with irl stuff. I actually updated [the readme](https://github.com/celestemods-com/CelesteMods/blob/main/public/images/mockups/modsPage/README.md) just a few hours ago with some more info, but the short version is: the mod doesn't get a `rating popover`, the maps do (we need to add a column to the maps table).

ShouvikGhosh2048 commented:
With the quality, difficulty and rate columns, the table will have seven columns. Should the carousel be shifted somewhere else?

ShouvikGhosh2048 commented:
@otobot1

The PR is ready for review.

ShouvikGhosh2048 commented on src/pages/mods/index.tsx:
```diff
@@ -21,26 +21,69 @@ const DEFAULT_PAGE_SIZE_INDEX = 1;
         },
+        expandedRow: {
+            // We move the expanded mod up to make it look like a single row.
+            "&&& + tr + tr": {
```
Styling the table is a bit tricky - margin/padding don't work on table rows, we don't set the `<td>` ourselves so we can't create an inner child and set padding on `<td>`. I went with setting `border-spacing` on the table which separates all rows equally, and then I shift the expanded mod up to make it look like a single row.

Let me know if there is a better way and I'll change this.

Reference: https://stackoverflow.com/questions/351058/space-between-two-rows-in-a-table

otobot1 replied:
Mantine DataTable exposes a [styles api](https://icflorescu.github.io/mantine-datatable/examples/additional-styling). It's been a while since I looked through it fully, but iirc the [multiple class names](https://icflorescu.github.io/mantine-datatable/examples/additional-styling#with-multiple-class-names) and [row styling](https://icflorescu.github.io/mantine-datatable/examples/additional-styling#row-styling) sections should have what you need.

ShouvikGhosh2048 replied:
I don't think this allows styling the expanded row. I can shift the translate to the `ExpandedMod` file, but the `translate` might be a bit confusing without the context that it is used in the mods page. I could also change `ExpandedMod` to take classes as a prop, and pass the `translate` there.

otobot1 replied:
Well the ExpandedMod component is only planned to be used on the mods page, so we can just put all of its styling in its file. If we wind up reusing it down the line we can refactor it then

otobot1 commented:
Sorry I havent had a chance to look through this yet. I'm in the middle of moving cities, so I probably wont get to it until next weekend...

ShouvikGhosh2048 commented:
I'll start working on other issues then.

otobot1 commented:
I finally had some time to quickly look at this. I haven't had a chance to review the code itself, but I have some styling notes:
1. The scrollbar overlaps the rows and header in the mods table. Maybe a few pixels of margin on the rows/header would solve the issue? If not, we will need to find another solution.
2. Please make the table footer the same colour as the table header - it kind of blends in to the background.
3. The rows in the mods table are quite tall. The individual cells in the table have 24 pixels of padding - please reduce this to 10px of top and bottom padding.
4. The rows have a funny interaction when they scroll up under the table header (see first screenshot). Please see if it looks better if the table header has square corners instead of rounded. If it still looks weird, then getting rid of the drop-shadow should help a little.
5. Thanks for getting the 1-click install component setup. It does take up quite a bit of vertical space though... Please reduce the font size of the installer button a little bit, and slightly increase the font size for publisher/publication date/more info so it looks more balanced.
6. Reduce the top/bottom padding within the Expanded Mod. The maps table title (`Maps`) and the row above it (see second screenshot) in particular should have less top/bottom padding.
7. The screenshot carousel is very small. Please change it so that it is at least twice the size.
![image](https://github.com/celestemods-com/CelesteMods/assets/27628434/53cc9d4c-6522-4b32-a9ae-91bbb95a45c5)
![image](https://github.com/celestemods-com/CelesteMods/assets/27628434/ddc43e4c-6ca9-4fdb-9d81-6ae38bb64bff)

Please let me know if anything is unclear. Thanks again for working on this!

ShouvikGhosh2048 commented:
@otobot1 

I've made the changes. For the footer, what should the color of the buttons be?

otobot1 commented:
> @otobot1
> 
> I've made the changes. For the footer, what should the color of the buttons be?

Good question! I'll play around with it and see what makes sense.

otobot1 commented:
> 1. The scrollbar overlaps the rows and header in the mods table. Maybe a few pixels of margin on the rows/header would solve the issue? If not, we will need to find another solution.

> 2. Please make the table footer the same color as the table header - it kind of blends in to the background.

> 3. The rows in the mods table are quite tall. The individual cells in the table have 24 pixels of padding - please reduce this to 10px of top and bottom padding.

> 4. The rows have a funny interaction when they scroll up under the table header (see screenshot). Please see if it looks better if the table header has square corners instead of rounded. If it still looks weird, then getting rid of the drop-shadow should help a little.

> 5. Thanks for getting the 1-click install component setup. It does take up quite a bit of vertical space though... Please reduce the font size of the installer button a little bit, and slightly increase the font size for publisher/publication date/more info so it looks more balanced.

> 6. Reduce the top/bottom padding within the Expanded Mod. The maps table title (`Maps`) and the row above it (see screenshot) in particular should have less top/bottom padding.

> 7. The screenshot carousel is very small. Please change it so that it is at least twice the size.

1. Thanks! Please also narrow the mods table footer so its right edge is in-line with the table header's right edge.

2. Turns out I'm actually not a fan of how that looks 😅. I'll adjust the footer color and set the button colors, hopefully tomorrow, or I'll make a follow-up issue for myself (TODO).

3. 👍

4. Looks much better! We should see if we can remove the drop-shadow until the table header is done sliding up (see screenshot).
![image](https://github.com/celestemods-com/CelesteMods/assets/27628434/01778047-7f69-48d9-98b9-b5f156903c39)

5. This still takes up too much vertical space... Maybe we should just change this to a single row of text that says "Olympus: 1-Click Install". I'll play with this tomorrow if I have a chance.

6. Much better, thanks! If possible, please also reduce the gap between the maps table title ("Maps") and the top of the maps table header.

7. Turns out 400px is too tall, but 250px seems fine. We also need to make the carousel width dynamic, as the image resolutions aren't guaranteed. We want to allow the carousel to grow vertically as needed if there is extra space because of the maps table, but we want to prevent it from growing past 250px unless there's at least 4 maps in the table. This can be a follow-up issue though so we can get this PR merged. Let me know what you prefer - it would still be part of v0.1.0.

ShouvikGhosh2048 commented:
@otobot1

For 1:
The table header lies inside the scroll area and can be larger than the scroll area's width while the footer lies outside the scroll area.

Normally the scroll area and footer have same width and the header is at least as wide as the scroll area width. I had added padding to the scroll area to prevent the scrollbar and header to overlap, which made the footer larger than the header.

If we make the footer smaller by adding margin, it becomes smaller than the scroll area, and so when the header is larger than the scroll area, it appears larger than the footer as seen below:

<img width="558" alt="celeste_mods_pagination" src="https://github.com/celestemods-com/CelesteMods/assets/91585022/e0636d78-ea5b-4ae2-91ed-39db63d1138c">

We could remove the padding from the scroll area - then even if the header is larger than the scroll area, the visible part of the header has the same width as the footer (the scroll bar will overlap with the header in this case):

<img width="564" alt="celeste_mods_pagination2" src="https://github.com/celestemods-com/CelesteMods/assets/91585022/fb974ce2-6909-441a-a5e7-d0b9a3b4be34">


For 4, I just removed the shadow below the header. I've made the changes for 6. I'd prefer to work on 7 as a separate pull request.

Any changes I need to make for 2 and 5?

otobot1 commented:
1. Would it be easier potentially to implement our own scrollbar instead of using the table's? We went with the built-in one because it was easier at the time, but if it's no longer easier... We can leave this as a follow-up issue if you agree.
2. Don't worry about this one. I figure it's easier if I sit down and play with the colors myself instead of going back and forth 20 times 😅 I'll make a follow-up issue for it.
4. 👍 
5. I modified the 1-Click Install to just be a single line - I think being compact is worth not having the recognizable button. I'll make a follow-up issue for making the 1-click install actually work.
6. 👍 
7. Sounds good. I'll make a follow-up issue.
